keps/74-support-argo-workflow/README.md

```diff
@@ -0,0 +1,722 @@
+
+    // 5. handle WaitForPodsReady only for a standalone job.
+    // handle a job when waitForPodsReady is enabled, and it is the main job
+
+    // 6. handle eviction
+
+	  // 7. handle job is suspended.
+
+	  // 8. handle job is unsuspended.
+    r.ReconcileStage(ctx, req, stage, wf)
+  }
+}
+```
+
+Each workflow managers can implement their own controller to integrate with the workflow framework.
+
+#### How to suspend by stages
+
+In my daily work, I primarily encounter workflow managers in the form of Argo and Tekton. Consequently, this section will focus on these two. Other workflow management systems can draw upon the discussions in this section for implementation guidance.
+
+These two workflow managers support to suspend a stage before the stage start to run. But they do
+not support to suspend a running stage (see https://github.com/tektoncd/pipeline/discussions/5353). 
+So only after the workflows support suspending a running stage, we can evict and reclaim the pod
+of the workflows. Eviction and reclain of stages can be implemented in the future. 
+
+We have three ways to suspend before starting a stage.
+
+Option 1. Give users a CLI to help users modifying workflows to add a specific suspend template for each step.
+When the workflows are suspended on this special suspend template, the job-controller in Kueue
+create workloads for the next step. Modification of workflow-controller is not needed for
+this way, so that it is easy to iterate, and no need to manage the version of argo and kueue.
+By in this way, users can modify their workflows to skip waiting in kueue, which maybe is not 
+acceptable for some users.
+
+![](suspend-with-cli.drawio.svg)
+
+For example, if users want to queue the workflow below:
+``` yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: loops-
+  namespace: argo
+spec:
+  entrypoint: loop-example-depth-1
+  templates:
+  - name: loop-example-depth-2
+    steps:
+    - - name: print-message-loop
+        template: print-message
+        arguments:
+          parameters:
+          - name: message
+            value: "{{item}}"
+        withItems:              # invoke print-message once for each item in parallel
+        - hello world           # item 1
+        - goodbye world         # item 2
+  - name: loop-example-depth-1
+    steps:
+    - - name: loop-example-depth-2
+        template: loop-example-depth-2
+        withItems:
+        - depth-1-1
+        - depth-1-2
+      - name: loop-example-depth-2-2
+        template: loop-example-depth-2
+  - name: print-message
+    inputs:
+      parameters:
+      - name: message
+    container:
+      image: busybox
+      command: [sleep]
+      args: ["3m"]
+```
+
+After they call the CLI, the new workflow will be like:
+``` yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: loops-
+  namespace: argo
+spec:
+  entrypoint: loop-example-depth-1
+  templates:
+  - name: kueue-suspend
```

Comment:
Does it mean that you are going to mutate the existing Workflow with `kueue-suspend` step ?
Why not just mutate the `suspend: true` parameter to every step ?

Reply:
> Does it mean that you are going to mutate the existing Workflow with kueue-suspend step ?

Yes

> Why not just mutate the suspend: true parameter to every step ?

Change suspend to `false` will start all suspend steps. Adding suspend template before every leaf template allow us to control the start of specific steps. For example, for the DAG like this:
```
A
|  \
B  C
```
I think step B and step C should wait for different workload admissions.